les/peer.go

```diff
@@ -1077,150 +1081,110 @@ type serverPeerSubscriber interface {
 // serverPeerSet represents the set of active server peers currently
 // participating in the Light Ethereum sub-protocol.
 type serverPeerSet struct {
-	peers map[string]*serverPeer
-	// subscribers is a batch of subscribers and peerset will notify
-	// these subscribers when the peerset changes(new server peer is
-	// added or removed)
-	subscribers []serverPeerSubscriber
-	closed      bool
-	lock        sync.RWMutex
+	ns     *nodestate.NodeStateMachine
+	opened uint32
+	closed bool
 }
 
 // newServerPeerSet creates a new peer set to track the active server peers.
-func newServerPeerSet() *serverPeerSet {
-	return &serverPeerSet{peers: make(map[string]*serverPeer)}
+func newServerPeerSet(ns *nodestate.NodeStateMachine) *serverPeerSet {
+	ns.SubscribeField(serverPeerField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if oldValue != nil {
+			oldValue.(*serverPeer).Peer.Disconnect(p2p.DiscUselessPeer)
+		}
+	})
+	return &serverPeerSet{ns: ns}
 }
 
 // subscribe adds a service to be notified about added or removed
 // peers and also register all active peers into the given service.
 func (ps *serverPeerSet) subscribe(sub serverPeerSubscriber) {
-	ps.lock.Lock()
```

Comment:
I think we are using the "node-state-machine" based peerset in the wrong way.
Essentially it's same with the map-based peerset. While the biggest difference is the nsm peerset supports subscription natively.

Originally, in order to support subscription, we need to implement the `registerPeer` and `unregisterPeer` for all the external components(downloader, fetcher, etc). It's annoying. So we should just get rid of these complexity.

In the new peerset, we should define these two methods:

```go
func (ps *serverPeerSet) OnRegister(callback func(node *enode.Node, state nodestate.Flags, peer *serverPeer)) {
	ps.ns.SubscribeField(serverPeerField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
		if newValue != nil {
			callback(node, state, newValue.(*serverPeer))
		}
	})
}

func (ps *serverPeerSet) OnUnregister(callback func(node *enode.Node, state nodestate.Flags, peer *serverPeer)) {
	ps.ns.SubscribeField(serverPeerField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
		if oldValue != nil {
			callback(node, state, oldValue.(*serverPeer))
		}
	})
}
``` 

So that all the subscribers only need to subscribe the peerset and define the relative actions

```go
peers.OnRegister(func(node *enode.Node, state nodestate.Flags, peer *serverPeer) {})
peers.OnUnregister(func(node *enode.Node, state nodestate.Flags, peer *serverPeer) {})
```

Reply:
It's the biggest benefit for this refracting I think(an elegant peerset subscription mechanism).